### PR TITLE
Do not store transitively required kernel libraries on disk.

### PR DESCRIPTION
**Pull request by ppedrot**
We only keep the root libraries both in the vo summary and in the dependency field of compiled libraries. This prevents a quadratic blowup of vo file size, since before each vo would keep the whole set of libraries it depends on, which is linear per vo file. Now the memory used by the set of vo files is linear in the size of the dependency DAG.

EDIT: doing this in the summary segment is tricky because Require relies on it and some objects are not idempotent, so this needs deeper fixes. For now I'm just tweaking the kernel part, it's easier.

### CODE REVIEW

**Comment by ppedrot** [comment]
@coqbot bench

**Comment by ppedrot** [comment]
Regardless of the buggy behaviour of extraction after this patch, that's a 10% reduction in vo size for the devs that compiled.

**Comment by Janno** [comment]
The change also reduces memory usage and reduces instruction count. This is awesome!

**Comment by Alizter** [comment]
What is going on with odd-order though?

**Comment by ppedrot** [comment]
@Alizter looks like noise to me.

**Comment by Alizter** [comment]
@ppedrot I wouldn't be so sure. It is not a small development and as far as developments go it is very linear since it is a series of files that sequentially depend on each other. I wonder if we are hitting some unoptimal behaviour here. Would you be able to run the bench again only on odd-order?

**Comment by silene** [comment]
We should really stop showing "user time" and "CPU cycles" as the first two columns. The bench machine can be running several benchmarks simultaneously, which greatly increases the memory pressure, thus causing data to be evicted from the caches (especially L3, but also L1 and L2 when the number of processes exceed the number of physical cores), and thus greatly degrading performances. Only the last two columns are relatively free from any outside interference.

That being said, the large memory increase on `coq-performance-tests-lite` is a bit worrying.

**Comment by ppedrot** [comment]
> That being said, the large memory increase on coq-performance-tests-lite is a bit worrying.

I thought we claimed that 16% increase in memory is the typical diff due to the GC kicking in at a different moment?

**Comment by ppedrot** [comment]
I've ended up performing the change only in the kernel data structure, because it's much trickier to implement this properly in the Library module. I should do that at some other time, but for now the easy tweak should go in.

**Comment by silene** [comment]
> I thought we claimed that 16% increase in memory is the typical diff due to the GC kicking in at a different moment?

To be more precise, a 20% increase in the major heap might be the GC kicking at different moment. But the last column does not show the increase of the major heap, it shows the increase of the whole memory fottprint (which includes the 256MB minor heap and various other data structures). So, the percentages have to be adjusted to take into account this immutable basis.

But in that case, it was my mistake, I miscounted the digits and messed my mental calculation. This 16% memory increase does indeed correspond to a 20% major heap increase.

**Comment by SkySkimmer** [comment]
>The bench machine can be running several benchmarks simultaneously

That should not be the case, do you have any examples of it happening?

**Comment by silene** [comment]
Yes, I have seen it happen several times. For example, compare the following two jobs. One finished on May 22th at 21:44; the other one started on May 22th at 19:36. So they overlap by two hours. https://gitlab.inria.fr/coq/coq/-/jobs/4350559 https://gitlab.inria.fr/coq/coq/-/jobs/4350670

**Comment by SkySkimmer** [comment]
Those are different machines, you can see on the right "Runner: #6459 (gGKko-ajq) ci-coq-04 (bench)" and "Runner: #6407 (v6HyzL39Q) ci-coq-03 (bench)"

**Comment by ppedrot** [comment]
Let's rerun a bench. This time it's unlikely we will see much improvement, most of the problems actually come from the library summary. @coqbot bench

**Comment by ppedrot** [comment]
I have submitted the library-level tweak in the separate PR #19128.

**Comment by ppedrot** [comment]
Leaderboard of the reductions:
```
coq-fiat-crypto-with-bedrock/./src/Everything.vo 103317 55210 48107 (-47%)
coq-fiat-crypto-with-bedrock/./src/PerfTesting/PerfTestPrint.vo 103445 52008 51437 (-50%)
coq-fiat-crypto-with-bedrock/./src/PerfTesting/PerfTestSearch.vo 103448 52011 51437 (-50%)
coq-fiat-crypto-with-bedrock/./src/PerfTesting/PerfTestSearchPattern.vo 103469 52032 51437 (-50%)
coq-fiat-crypto-with-bedrock/./src/Bedrock/End2End/X25519/EdwardsXYZT.vo 1619587 1567407 52180 (-3%)
coq-fiat-crypto-with-bedrock/./src/Bedrock/End2End/X25519/MontgomeryLadder.vo 507129 454802 52327 (-10%)
coq-fiat-crypto-with-bedrock/./src/Bedrock/End2End/X25519/MontgomeryLadderRISCV.vo 345315 292352 52963 (-15%)
coq-fiat-crypto-with-bedrock/./src/Bedrock/End2End/X25519/GarageDoor.vo 3120405 3067050 53355 (-2%)
coq-fiat-crypto-with-bedrock/./src/Bedrock/End2End/X25519/GarageDoorTop.vo 632515 578881 53634 (-8%)
coq-fiat-crypto-with-bedrock/./src/Bedrock/Everything.vo 136780 72379 64401 (-47%)
TOTAL 1148442675 1091523696 -56918979 (-5%)
```

File: kernel/safe_typing.ml [hunk 1]
```diff
@@ -493,16 +498,21 @@ let check_empty_struct senv =
...
+  let check current (id, required) = match DPmap.find_opt id current with
+  | None ->
+    CErrors.user_err Pp.(pr_sequence str ["Reference to unknown module"; DirPath.to_string id; "."])
+  | Some { req_root; req_digest = actual } ->
+    if not (digest_match ~actual ~required) then
+      CErrors.user_err Pp.(pr_sequence str
+        ["Inconsistent assumptions over module"; DirPath.to_string id; "."])
+    else if req_root then
+      (* the library is being transitively required, not a root anymore *)
+      DPmap.add id { req_root = false; req_digest = actual } current
```
Comment (SkySkimmer) [inline kernel/safe_typing.ml:510]:
```suggestion
      DPmap.set id { req_root = false; req_digest = actual } current
```

Comment (ppedrot) [reply kernel/safe_typing.ml]: Nitpick addressed.

**Review by SkySkimmer — APPROVED** [review]

**Comment by SkySkimmer** [comment]
@coqbot merge now

**Pull request merged**